File: 1675-minimize-deviation-in-array/python3/1675-minimize-deviation-in-array.py
import heapq

class Solution:
    def minimumDeviation(self, nums: List[int]) -> int:
      pq = [-(x if x % 2 == 0 else x * 2) for x in nums]
      heapq.heapify(pq)
      min_v = -max(pq)
      min_dev = float("inf")
      while True:
        max_v = -heapq.heappop(pq)
        min_dev = min(min_dev, max_v - min_v)
        if max_v % 2 == 1:
          break
        max_v //= 2
        min_v = min(min_v, max_v)
        heapq.heappush(pq, -max_v)
      return min_dev

File: 1675-minimize-deviation-in-array/java/1675-minimize-deviation-in-array.java
class Solution {
  public int minimumDeviation(int[] nums) {
    var pq = new PriorityQueue<Integer>(Comparator.<Integer>comparingInt(v -> v).reversed());

    int min_v = Integer.MAX_VALUE;
    int min_dev = Integer.MAX_VALUE;

    for (var i = 0; i < nums.length; ++i) {
      var v = nums[i];
      if (v % 2 == 1) {
        v *= 2;
      }
      min_v = Math.min(min_v, v);

      pq.add(v);
    }

    while (true) {
      var max_v = pq.poll();
      min_dev = Math.min(min_dev, max_v - min_v);
      if (max_v % 2 == 1) {
        break;
      }
      max_v /= 2;
      min_v = Math.min(min_v, max_v);
      pq.add(max_v);
    }

    return min_dev;
  }
}

